x/orderbook/controller.go

```diff
@@ -0,0 +1,166 @@
+	}
+	// figure out price and if we want highest or lowest
+	matches, err := c.orderBucket.PrefixScan(db, prefix, descending)
+	if err != nil {
+		return errors.Wrap(err, "prefix scan")
+	}
+
+	// process every match until the offer is closed
+	for order.RemainingOffer.IsPositive() && matches.Valid() {
+		var match Order
+		err := matches.Load(&match)
+		if err != nil {
+			return errors.Wrap(err, "loading match")
+		}
+
+		// if price doesn't match, break out of loop
+		if !acceptable(order.Price, match.Price) {
+			break
+		}
+
+		// otherwise, execute trade
+		err = c.executeTrade(db, order, &match, now)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// execute trade assumes this was already validated as acceptable.
+// it uses the price of the counter offer (which is equal to or better
+// that what the new order requested).
+// the smaller order will be emptied, and the larger one typically
+// left with some remaining balance
+func (c controller) executeTrade(db weave.KVStore, taker, maker *Order, now weave.UnixTime) error {
+	ask, bid := taker, maker
+	if taker.Side == Side_Bid {
+		bid, ask = taker, maker
+	}
+
+	askVal, bidVal, err := amountToSettle(ask, bid, maker.Price)
+	if err != nil {
+		return errors.Wrapf(err, "executing trade %X with %X", taker.ID, maker.ID)
+	}
+
+	takerPaid, makerPaid := askVal, bidVal
+	if taker.Side == Side_Bid {
+		makerPaid, takerPaid = askVal, bidVal
+	}
+
+	// create a trade record
+	trade := Trade{
+		Metadata:    &weave.Metadata{},
+		OrderBookID: ask.OrderBookID,
+		TakerID:     taker.ID,
+		MakerID:     maker.ID,
+		Taker:       taker.Trader,
+		Maker:       maker.Trader,
+		TakerPaid:   &takerPaid,
+		MakerPaid:   &makerPaid,
+		ExecutedAt:  now,
+	}
+	if err := c.tradeBucket.Put(db, &trade); err != nil {
+		return errors.Wrap(err, "saving trade")
+	}
+
+	if err := c.payout(db, ask, bid.Trader, askVal, now); err != nil {
+		return errors.Wrap(err, "ask payout")
+	}
+	if err := c.payout(db, bid, ask.Trader, bidVal, now); err != nil {
+		return errors.Wrap(err, "bid payout")
+	}
+
+	return nil
+}
+
+func (c controller) payout(db weave.KVStore, from *Order, to weave.Address, amount coin.Coin, now weave.UnixTime) error {
+	// payout the ask side
+	err := c.mover.MoveCoins(db, from.Address(), to, amount)
+	if err != nil {
+		return errors.Wrap(err, "paying trader")
+	}
+	rem, err := from.RemainingOffer.Subtract(amount)
+	if err != nil {
+		return errors.Wrap(err, "deducting ask tokens")
+	}
+	from.RemainingOffer = &rem
+	from.UpdatedAt = now
+	if from.RemainingOffer.IsZero() {
+		// TODO: remember the hanging chads (rr... 1 fractional)
+		from.OrderState = OrderState_Done
+	}
+	if err := c.orderBucket.Put(db, from); err != nil {
+		return errors.Wrap(err, "updating order")
+	}
+	return nil
+}
+
+func amountToSettle(ask *Order, bid *Order, price *Amount) (askVal, bidVal coin.Coin, err error) {
+	// TODO: how to handler remainders and rounding?
```

Comment:
Usually on traditional centralized exchanges round down logic is used. But it is intentional. Even those small left over fractions makes credible amount of money after thousands of trades. Check [peatio/fee_chargeable.rb](https://github.com/rubykube/peatio/blob/master/app/models/concerns/fee_chargeable.rb#L30), [peatio/market.rb](https://github.com/rubykube/peatio/blob/master/app/models/market.rb#L89). It seems there are better solutions to this https://bitcoin.stackexchange.com/a/36938

Reply:
This is a good point, and a TODO that should be resolved before merge.

Point is that I offer a bid with a price X. After trading y ask for X-delta bid tokens, I have some small amount (delta) left in the order. That value is less than 1 fractional ask token * price, and thus could never be deducted. 

This order will remain in the orderbook forever, loaded on all matches, and settled with no change. It will not break logic (I think), but as they accumulate, they will slow down the machinery.

It would be good to figure out the business logic here. I guess taking the last fraction and sending them to the market owner is typical? Or returning them to the order creator is most fair?

Reply:
What about seperating that logic as a function. So users of orderbook module can desing which fractional left over strategy. But for now we can use the simplest one: return it to order maker.

Reply:
Sure, we need to define a function type, and maybe two sample implementations. To be set in the constructor.

Good idea.

Reply:
👍 good discussion and good outcome